contentcuration/contentcuration/frontend/channelEdit/vuex/clipboard/getters.js

```diff
@@ -0,0 +1,332 @@
+   *
+   * @param {string} id
+   */
+  return function(id) {
+    const rootId = rootGetters['clipboardRootId'];
+
+    if (id === rootId) {
+      return null;
+    }
+
+    if (getters.channelIds.includes(id)) {
+      return rootId;
+    }
+
+    const treeNode = rootGetters['contentNode/getTreeNode'](id);
+
+    if (!treeNode) {
+      return null;
+    }
+
+    return treeNode.parent === rootId ? treeNode.channel_id : treeNode.parent;
+  };
+}
+
+/**
+ * List of distinct source channel ID's containing a node on the clipboard
+ */
+export function channelIds(state, getters) {
+  return uniq(getters.clipboardChildren.map(n => n.channel_id));
+}
+
+/**
+ * List of distinct source channel containing a node on the clipboard
+ */
+export function channels(state, getters, rootState, rootGetters) {
+  return getters.channelIds.map(id => rootGetters['channel/getChannel'](id));
+}
+
+export function filterSelectionIds(state) {
+  /**
+   * Filters the selection states by a flag|state
+   *
+   * @param {Number|Function} filter
+   */
+  return function(filter) {
+    const filterFunc = isFunction(filter)
+      ? filter
+      : (id, selectionState) => Boolean(selectionState & filter);
+
+    return Object.entries(state.selected)
+      .map(([id, selectionState]) => (filterFunc(id, selectionState) ? id : null))
+      .filter(Boolean);
+  };
+}
+
+/**
+ * List of the selected tree nodes on the clipboard
+ */
+export function selectedNodes(state, getters, rootState, rootGetters) {
+  return getters
+    .filterSelectionIds(SelectionFlags.SELECTED)
+    .map(id => rootGetters['contentNode/getTreeNode'](id))
+    .filter(Boolean);
+}
+
+/**
+ * List of channels containing a selected node on the clipboard
+ */
+export function selectedChannels(state, getters, rootState, rootGetters) {
+  return getters.selectedNodes
+    .map(node => node.channel_id)
+    .filter(Boolean)
+    .reduce((channelIds, channelId) => {
+      if (!channelIds.includes(channelId)) {
+        channelIds.push(channelId);
+      }
+
+      return channelIds;
+    }, [])
+    .map(channelId => rootGetters['channel/getChannel'](channelId));
+}
+
+export function getChannelColor(state) {
+  /**
+   * The visual color cue for the channel, determined from the thumbnail
+   *
+   * @param {string} channelId
+   * @param {string} [defaultValue='#6c939b']
+   */
+  return function(channelId, defaultValue = '#6c939b') {
+    return channelId in state.channelColors ? state.channelColors[channelId] : defaultValue;
+  };
+}
+
+export function currentSelectionState(state) {
+  /**
+   * The current selection state for a node as it is in the state
+   *
+   * @param {string} id
+   */
+  return function(id) {
+    return state.selected[id];
+  };
+}
+
+export function getSelectionState(state, getters, rootState, rootGetters) {
+  /**
+   * Get the current computed state of the node. The state is a bitmask of the
+   * selection flags
+   *
+   * All the states:
```

Comment:
Could you maybe provide some examples of what might fall under each of these states? Maybe something like:
```
3 = Node is selected, but indeterminate (e.g.  [-] Topic -> [x] Node means copy all descendants, but not topic)
4 = All descendants are selected (e.g. [x] Topic -> [x] Node  means copy topic and all its descendants)
```

Reply:
I think it would be easier to understand the states by looking at the flags. Perhaps listing these states here is too confusing? I say that because state `4 != "copy topic and all its descendants"`. This is helping me understand what's not understandable :smile: 

I think it might help if you think of the selection state as solely what you need to know in order to render the checkbox for that item.
1. If the node is *not* a topic node, the only flag we care about is the `SELECTED` flag
2. If the node is a topic node, we need to know:
  a. if the node itself is selected, which again is the `SELECTED` flag,
  b. whether there exists a selected descendant node, represented by `INDETERMINATE`, and
  c. whether all the descendants are selected through `ALL_DESCENDANTS`.

Where the selection flags are:
```
  NONE: 0,
  SELECTED: 1,
  INDETERMINATE: 2,
  ALL_DESCENDANTS: 4,
```

So taking your example, it would be:
```
State 1 = 1 = SELECTED
  e.g. [X] Node State 1
State 2 = 2 = INDETERMINATE
  e.g. [-] Channel State 2 -> { [ ] Node A, [X] Node B }
State 3 = 1 & 2 = SELECTED & INDETERMINATE
  e.g. [X] Topic State 3 -> { [ ] Node A, [X] Node B }
State 4 = 4 = ALL_DESCENDANTS
  e.g. [X] Channel State 4
State 5 = 1 & 4 = SELECTED & ALL_DESCENDANTS
  e.g. [X] Topic State 5 -> { [ ] Node A, [X] Node B }
State 6 = 2 & 4 = INDETERMINATE & ALL_DESCENDANTS
  e.g. [ ] Topic State 6 -> { [X] Node A, [X] Node B }
```

Does this help? Should I put this in the comments?

Reply:
That's perfect! Yeah, definitely makes it a little easier to understand the differences